### PR TITLE
fix: Handle exception in `validate`

### PR DESCRIPTION
**Pull request by sqren**
Closes https://github.com/enquirer/enquirer/issues/277

Instead of causing unhandled promise rejections, exceptions in `validate` should cause the prompt to end and the exception bubble up.

Now exceptions are caught:
```
✔ What is your username? · John Doe
An error occurred and the program ended :(
```

Instead of being unhandled:
```
? What is your username? › John Doe (node:9358) UnhandledPromiseRejectionWarning: Error: boo!
    at Input.validate (/Users/sqren/srv/enquirer-demo/index.js:8:11)
    at Input.submit (/Users/sqren/srv/enquirer-demo/node_modules/enquirer/lib/prompt.js:123:49)
    at Input.submit (/Users/sqren/srv/enquirer-demo/node_modules/enquirer/lib/prompts/input.js:51:18)
    at Input.keypress (/Users/sqren/srv/enquirer-demo/node_modules/enquirer/lib/prompt.js:44:23)
    at Input.keypress (/Users/sqren/srv/enquirer-demo/node_modules/enquirer/lib/types/string.js:24:18)
    at ReadStream.on (/Users/sqren/srv/enquirer-demo/node_modules/enquirer/lib/keypress.js:205:26)
    at ReadStream.emit (events.js:203:15)
    at emitKeys (internal/readline.js:424:14)
    at emitKeys.next (<anonymous>)
    at ReadStream.onData (readline.js:1073:36)
(node:9358) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:9358) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

### CODE REVIEW

File: lib/prompt.js [hunk 1]
```diff
@@ -120,7 +120,18 @@ class Prompt extends Events {
...
+    let result;
+    try {
+      result =
+        this.state.error || (await this.validate(this.value, this.state));
+    } catch (e) {
+      this.state.cancelled = true;
+      await this.render();
+      await this.close();
+      this.emit("cancel", e);
+      return;
```
Comment (sqren) [inline lib/prompt.js:132]: I wanted to replace these lines with `this.cancel(e)` however due to the logic in `this.error(err)` the error is not propagated correctly.

**Comment by sqren** [comment]
@jonschlinkert @doowb Sorry for the direct pings 🙉 
 Anything I can do on my end to get this moving?
Thanks! 🙇‍♂️

**Comment by sqren** [comment]
Seems a little quiet in here :) Is the project no longer actively maintained?

File: lib/prompt.js [hunk 2]
```diff
@@ -120,7 +120,18 @@ class Prompt extends Events {
       await this.options.onSubmit.call(this, this.name, this.value, this);
     }
 
-    let result = this.state.error || await this.validate(this.value, this.state);
+    let result;
+    try {
+      result =
+        this.state.error || (await this.validate(this.value, this.state));
```
Comment (jonschlinkert) [inline lib/prompt.js:126]: please format this to be on one line. thanks!

Comment (sqren) [reply lib/prompt.js]: Done

**Pull request closed without merging**